pkg/scheduler/admission/podnodeconstraints/admission.go

```diff
@@ -0,0 +1,77 @@
+package podnodeconstraints
+
+import (
+	"fmt"
+	"io"
+	"io/ioutil"
+	"reflect"
+
+	configlatest "github.com/openshift/origin/pkg/cmd/server/api/latest"
+	"github.com/openshift/origin/pkg/scheduler/admission/podnodeconstraints/api"
+
+	admission "k8s.io/kubernetes/pkg/admission"
+	kapi "k8s.io/kubernetes/pkg/api"
+	kclient "k8s.io/kubernetes/pkg/client/unversioned"
+)
+
+func init() {
+	admission.RegisterPlugin("PodNodeConstraints", func(c kclient.Interface, config io.Reader) (admission.Interface, error) {
+		pluginConfig, err := readConfig(config)
+		if err != nil {
+			return nil, err
+		}
+		return NewPodNodeConstraints(pluginConfig), nil
+	})
+}
+
+func NewPodNodeConstraints(config *api.PodNodeConstraintsConfig) admission.Interface {
+	return &podNodeConstraints{
+		config:  config,
+		Handler: admission.NewHandler(admission.Create),
+	}
+}
+
+type podNodeConstraints struct {
+	*admission.Handler
+	config *api.PodNodeConstraintsConfig
+}
+
+func readConfig(reader io.Reader) (*api.PodNodeConstraintsConfig, error) {
+	if reader == nil || reflect.ValueOf(reader).IsNil() {
+		return &api.PodNodeConstraintsConfig{}, nil
+	}
+
+	configBytes, err := ioutil.ReadAll(reader)
+	if err != nil {
+		return nil, err
+	}
+	config := &api.PodNodeConstraintsConfig{}
+	err = configlatest.ReadYAML(configBytes, config)
+	if err != nil {
+		return nil, err
+	}
+	// No validation needed since config is just 2 bools
+	return config, nil
+}
+
+func (oo *podNodeConstraints) Admit(aa admission.Attributes) error {
+	switch {
+	case oo.config == nil,
+		aa.GetResource() != kapi.Resource("pods"),
+		aa.GetKind().Kind != "Pod",
+		aa.GetSubresource() != "",
+		aa.GetOperation() != admission.Create:
+		return nil
+	}
+	pod, ok := aa.GetObject().(*kapi.Pod)
+	if !ok {
+		return admission.NewForbidden(aa, fmt.Errorf("unexpected object: %#v", aa.GetObject()))
+	}
+	if len(pod.Spec.NodeName) > 0 && oo.config.ProhibitNodeName {
```

Comment:
On Thu, Feb 18, 2016 at 5:01 PM, Dan McPherson <notifications@github.com>
wrote:

> In pkg/scheduler/admission/podnodeconstraints/admission.go
> <https://github.com/openshift/origin/pull/6979#discussion_r53389734>:
>
> > +}
> > +
> > +func (oo *podNodeConstraints) Admit(aa admission.Attributes) error {
> > +	switch {
> > +	case oo.config == nil,
> > +		aa.GetResource() != kapi.Resource("pods"),
> > +		aa.GetKind().Kind != "Pod",
> > +		aa.GetSubresource() != "",
> > +		aa.GetOperation() != admission.Create:
> > +		return nil
> > +	}
> > +	pod, ok := aa.GetObject().(*kapi.Pod)
> > +	if !ok {
> > +		return admission.NewForbidden(aa, fmt.Errorf("unexpected object: %#v", aa.GetObject()))
> > +	}
> > +	if len(pod.Spec.NodeName) > 0 && oo.config.ProhibitNodeName {
>
> @smarterclayton <https://github.com/smarterclayton>
>
> To the first one I don't find it convincing. Please describe a
> scenario where anything that allows escape of a container is anything
> less than access to other nodes. The virus example is weak - I can
> think of nothing that would actually stop reaching 100% saturation,
> including this PR.
>
> Regardless of whether we can think of one. You are giving someone a way to
> get to every node. Even if they have nothing but read access to the other
> containers. That's an extra hole.
>

It's not meaningful.  It doesn't justify a particular change.


> The second argument is weak as well - a node that's almost full but
> has low CPU activity. So by adding their preferential location, they
> are increasing utilization of a resource that would otherwise be
> wasted. What's the alternative? They sit on another node and use the
> exact same amount of CPU they would use before? They just did us a
> favor.
>
> There are multiple ways this isn't doing a favor:
>
>    - Gives everyone on those nodes who have occasional usage a worse
>    experience
>
>
Only when there is a small bucket left (such that a pod could be scheduled
there).  If you're arguing that we shouldn't allow pods to get onto nodes
that are *almost* full, then that's not really the point of this PR and it
won't do anything to improve that.


>
>    - Messes up utilization data on those machines which might have fit
>    into a lower utilization cost bucket
>
>
Not sure what you mean here.


>
>    - Remove the possibility of giving preferential treatment to other
>    pods by placing them on these sorts of nodes
>
> That's the rescheduler job.


>
>
> Beyond those reasons this removes the ability of the scheduler and
> operators from controlling the systems and provides a users with an
> experience that isn't desired by operators or the platform providers.
>
The role of preventing abuse generally starts by addressing the most
important abuse problem first, then the next, then the next.  Arguing this
particular change is important neglects the other more important abuse
scenarios that it doesn't fix.  We laid this out today, but my
recommendation was not to do pods/bind, it was to actually go and fix the
underlying issues

1. The kubelet shouldn't see pods that fail constraints - in general the
node has too high privileges today and sees all pods.  It needs to see only
the pods it has been scheduled (is entitled to see).  And any logic that
solves that can also integrate the "this is an impossible match" and filter
it out, which prevents the pod from being affected
2. It's still possible with this change to abuse nodes directly via label
selector and fast rescheduling, or just with fast rescheduling.  The
problem is rate limiting of pods per namespace, not preventing node name.

Addressing those fixes actual problems.  However, in practice, an operator
can simply monitor for "rejected due to constraints" and go ban that user.
I reject in practice the "target nodes is our security problem", because in
practice all dangerous problems are node escape problems, and this doesn't
address those.